dlib/dnn/layers.h

```diff
@@ -2463,11 +2492,105 @@ namespace dlib
         resizable_tensor params, empty_params; 
         alias_tensor gamma, beta;
         layer_mode mode;
+        bool is_disabled = false;
     };
 
     template <typename SUBNET>
     using affine = add_layer<affine_, SUBNET>;
 
+    namespace impl
+    {
+        class visitor_fuse_convolutions
+        {
+            public:
+            template <typename T> void fuse_convolutions(T&) const
+            {
+                // disable other layer types
+            }
+
+            // handle the standard case (convolutional layer followed by affine;
+            template <long nf, long nr, long nc, int sy, int sx, int py, int px, typename U, typename E>
+            void fuse_convolutions(add_layer<affine_, add_layer<con_<nf, nr, nc, sy, sx, py, px>, U>, E>& l)
+            {
+                // get the parameters from the affine layer as alias_tensor_instance
+                auto gamma = l.layer_details().get_gamma();
+                auto beta = l.layer_details().get_beta();
+
+                // get the convolution below the affine layer and its paramaters
+                auto& conv = l.subnet().layer_details();
+                const long num_filters_out = conv.num_filters();
+                const long num_rows = conv.nr();
+                const long num_cols = conv.nc();
+                tensor& params = conv.get_layer_params();
+                // guess the number of input filters
+                long num_filters_in;
+                if (conv.bias_is_disabled())
+                    num_filters_in = params.size() / num_filters_out / num_rows / num_cols;
+                else
+                    num_filters_in = (params.size() - num_filters_out) / num_filters_out / num_rows / num_cols;
+
+                // set the new number of parameters for this convolution and enable bias if needed
+                const size_t num_params = num_filters_in * num_filters_out * num_rows * num_cols + num_filters_out;
+                alias_tensor filters(num_filters_out, num_filters_in, num_rows, num_cols);
+                alias_tensor biases(1, num_filters_out);
+                if (conv.bias_is_disabled())
+                {
+                    conv.enable_bias();
+                    resizable_tensor new_params = params;
+                    new_params.set_size(num_params);
+                    biases(new_params, filters.size()) = 0;
+                    params = new_params;
+                }
```

Comment:
I've realized this block does not work properly. I get a segfault... I guess it's not a good idea to resize the params of the layer once it has already been allocated. What would be a nice way of doing it?
Anyway, if the user is planning on fusing the convolutions after the training, it shouldn't matter that much, but still...